quipucords/scanner/input_log.py

```diff
@@ -0,0 +1,301 @@
+# Copyright (c) 2018 Red Hat, Inc.
+#
+# This software is licensed to you under the GNU General Public License,
+# version 3 (GPLv3). There is NO WARRANTY for this software, express or
+# implied, including the implied warranties of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv3
+# along with this software; if not, see
+# https://www.gnu.org/licenses/gpl-3.0.txt.
+
+"""Log all input to the filesystem for later debugging."""
+
+import json
+import logging
+import multiprocessing as mp
+import os
+import time
+import uuid
+from django.db import transaction
+from api import log_state
+from quipucords import settings
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+UUID_CACHE = None
+
+DEFAULT_MAX_SIZE = 1 << 30  # 1 GiB
+DEFAULT_MAX_AGE = 365 * 24 * 60 * 60  # 1 year, except on leap years
+
+
+@transaction.atomic()
+def get_sonar_uuid():
+    """Get the UUID of this Sonar installation."""
+    global UUID_CACHE  # pylint: disable=global-statement
+    if UUID_CACHE is not None:
+        return UUID_CACHE
+
+    uuid_obj = log_state.DatabaseUUID.objects.all().first()
+    if uuid_obj is None:
+        uuid_obj = log_state.DatabaseUUID(uuid=uuid.uuid4())
+        uuid_obj.save()
+    UUID_CACHE = uuid_obj.uuid
+
+    return uuid_obj.uuid
+
+
+# This is a potential performance bottleneck since we are touching the
+# database every time we write a record. This isn't necessary - we
+# could get the state of the counter by reading the last few entries
+# of the existing log files on startup. I don't plan to implement that
+# unless we find that this is actually a problem.
+@transaction.atomic()
+def next_sequence_number():
+    """Get the next sequence number and increment the counter."""
+    seq = log_state.LatestSequenceNumber.objects.all().first()
+    if seq is None:
+        seq = log_state.LatestSequenceNumber(number=0)
+        seq.save()
+
+    number = seq.number
+    seq.number += 1
+    seq.save()
+
+    return number
+
+
+def find_prefix_scan_stat(prefix, dirname):
+    """Find files in a directory matching a prefix, and stat them.
+
+    This function exists because os.scandir isn't in Python 3.4, which
+    we support.
+
+    :param prefix: the name prefix to search for.
+    :param dirname: the directory to scan.
+
+    :returns: a list of (name, stat) pairs for matching files.
+    """
+    result = []
+
+    for name in os.listdir(dirname):
+        if name.startswith(prefix):
+            result.append((name, os.stat(os.path.join(dirname, name))))
+
+    return result
+
+
+# pylint: disable=too-many-instance-attributes
+class RotatingLogFile(object):
```

Comment:
Feel like we could have used `RotatingFileHandler`, instead of inventing our own.
I think we care more about the size of the data then whether it ages out after a year. How much data do we generate for a single system today?
I'm interested in understanding what we think an average amount would be, so we know when we'd hit a GB.

Reply:
I looked at `RotatingFileHandler`, but it seems to be very tied to the `logging` module's API where there are a record class and a formatter class that work together with the handler. We would have to write a custom formatter to get the output we want, and package up our output in a way that is not really what the logging module expects. I can do it that way if you want, it just seemed simpler to write this myself. (https://github.com/python/cpython/blob/3.6/Lib/logging/__init__.py)

Reply:
It just might solve many of the the flaws you pointed to below and is well tested.